File: rvc_service/RVCSDK/examples/CSharp/SaveLoadCaptureOptions.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RVC_CSharp;

namespace RVC_CSharp
{
    class Program
    {
        static void UseX1()
        {
            #region Step 0 ,Init & Find & Open

            if (false == RVC_CSharp.System.Init())
            {
                Console.WriteLine("Failed to init system.");
                Console.ReadKey();
                return;
            }

            Device device = RVC_CSharp.System.FindDeviceByIndex(0);
            if (false == device.IsValid() || false == device.CheckCanConnected())
            {
                Console.WriteLine("Device Can not be Connected.\nPlease Check Connection.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            // optional 
            // Print Device Info
            DeviceInfo info = new DeviceInfo();
            device.GetDeviceInfo(ref info);
            info.Print();

            if (info.support_x1 == false)
            {
                Console.WriteLine("Device not support x1!");
                Console.ReadKey();
                return;
            }
            if(device.IsFirmwareMatch() == false)
            {
                Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            X1 camera = RVC_CSharp.X1.Create(device, CameraID.CameraID_Left);
            camera.Open();
            if (false == camera.IsValid() || false == camera.IsOpen())
            {
                Console.WriteLine($"Failed to open camera .Error: {System.GetLastErrorMessage()} ");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }


            #endregion

            #region Step 1 ,[Main Step] Modify Options And Capture

            X1.CaptureOptions cap_opt = X1.CaptureOptions.Default();

            // Transform point map's coordinate to camera or reference plane
            cap_opt.transform_to_camera = true;
            // Set noise points filter range (0~30)
            cap_opt.filter_range = 0;
            // Set projector brightness (1~240)
            cap_opt.projector_brightness = 200;
            // Set camera exposure time (3~100) ms
            cap_opt.exposure_time_2d = 20;
            cap_opt.exposure_time_3d = 20;

            // Set HDR exposure times [0, 2, 3]. 0 presents not use hdr. 2 and 3 presents hdr times.
            cap_opt.hdr_exposure_times = 3;
            cap_opt.hdr_exposuretime_content = new int[3];
            cap_opt.hdr_exposuretime_content[0] = 3;
            cap_opt.hdr_exposuretime_content[1] = 6;
            cap_opt.hdr_exposuretime_content[2] = 12;

            bool ret = false;

            ret = camera.SaveCaptureOptionParameters(cap_opt);
            if (ret)
            {
                Console.WriteLine("Save Capture Option Parameters Succeed");
            }
            else
            {
                Console.WriteLine($"Save Capture Option Parameters Failed. Error: {System.GetLastErrorMessage()} ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.WriteLine("Press any key to exit. ");
                Console.ReadKey();
                return;
            }

            X1.CaptureOptions options = new X1.CaptureOptions();

            ret = camera.LoadCaptureOptionParameters(ref options);
            if (ret)
            {
                Console.WriteLine("Load Capture Option Parameters Succeed");
            }
            else
            {
                Console.WriteLine($"Load Capture Option Parameters Failed. Error: {System.GetLastErrorMessage()} ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.WriteLine("Press any key to exit. ");
                Console.ReadKey();
                return;
            }

            options.Print();
            #endregion

            #region Step 2 , Close

            camera.Close();
            camera.Destroy();

            RVC_CSharp.System.Shutdown();
            Console.WriteLine("System close.");

            Console.ReadKey();
            #endregion
            return;

        }
        static void UseX2()
        {
            #region Step 0 ,Init & Find & Open

            if (false == RVC_CSharp.System.Init())
            {
                Console.WriteLine("Failed to init system.");
                return;
            }

            Device device = RVC_CSharp.System.FindDeviceByIndex(0);
            //Device device = RVC_CSharp.System.FindDeviceBySN("P2GM353W002");
            if(device.IsFirmwareMatch() == false)
            {
                Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            
            if (false == device.IsValid() || false == device.CheckCanConnected())
            {
                Console.WriteLine("Device Can not be Connected.\nPlease Check Connection.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            X2 camera = RVC_CSharp.X2.Create(device);
            camera.Open();
            if (false == camera.IsValid() || false == camera.IsOpen())
            {
                Console.WriteLine($"Failed to open camera .Error: {System.GetLastErrorMessage()} ");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            // optional 
            // Print Device Info
            DeviceInfo info = new DeviceInfo();
            device.GetDeviceInfo(ref info);
            info.Print();
            #endregion

            #region Step 1 ,[Main Step] Modify Options And Capture

            X2.CaptureOptions cap_opt = X2.CaptureOptions.Default();

            // Transform point map's coordinate to camera or reference plane
            cap_opt.transform_to_camera = RVC_CSharp.CameraID.CameraID_NONE;
            // Set light contrast threshold range (0~10). the default value is 3.
            cap_opt.light_contrast_threshold = 3;
            // Set projector brightness (1~240)
            cap_opt.projector_brightness = 200;
            // Set camera exposure time (3~100) ms
            cap_opt.exposure_time_2d = 20;
            cap_opt.exposure_time_3d = 20;
            // Set projector color
            cap_opt.projector_color = RVC_CSharp.ProjectorColor.ProjectorColor_Blue;

            bool ret = false;

            ret = camera.SaveCaptureOptionParameters(cap_opt);

            if (ret)
            {
                Console.WriteLine("Save Capture Option Parameters Succeed");
            }
            else
            {
                Console.WriteLine($"Save Capture Option Parameters Failed. Error: {System.GetLastErrorMessage()} ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.WriteLine("Press any key to exit. ");
                Console.ReadKey();
                return;
            }

            X2.CaptureOptions options = new X2.CaptureOptions();

            ret = camera.LoadCaptureOptionParameters(ref options);

            if (ret)
            {
                Console.WriteLine("Load Capture Option Parameters Succeed");
            }
            else
            {
                Console.WriteLine($"Load Capture Option Parameters Failed. Error: {System.GetLastErrorMessage()} ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.WriteLine("Press any key to exit. ");
                Console.ReadKey();
                return;
            }


            options.Print();

            #endregion

            #region Step 2 ,  Close

            camera.Close();
            camera.Destroy();

            RVC_CSharp.System.Shutdown();
            Console.WriteLine("System close.");

            Console.ReadKey();
            #endregion

            return;
        }
        static void Main()
        {
            UseX1();
            UseX2();
            Console.ReadKey();
            return;
        }

    }
}

File: rvc_service/RVCSDK/examples/Python/SaveLoadCaptureOptions.py
# Copyright (c) RVBUST, Inc - All rights reserved.
import PyRVC as RVC
import os
import numpy as np
from Utils.Tools import *

def UseX1():
    # Initialize RVC system.
    RVC.SystemInit()

    # Choose RVC Camera type (USB, GigE or All)
    opt = RVC.SystemListDeviceTypeEnum.All

    # Scan all RVC Camera devices.
    ret, devices = RVC.SystemListDevices(opt)
    print("RVC Camera devices number:%d" % len(devices))

    # Find whether any RVC Camera is connected or not.
    if len(devices) == 0:
        print("Can not find any RVC Camera!")
        RVC.SystemShutdown()
        return 1
    print("devices size = %d" % len(devices))
    if devices[0].IsFirmwareMatch() == False:
        print("device firmware mismatch, Please use RVCManager to upgrade the firmware")
        RVC.SystemShutdown()
        return 1
    # Create a RVC Camera and choose use left side camera.
    x = RVC.X1.Create(devices[0], RVC.CameraID_Left)

    # Test RVC Camera is valid or not.
    if x.IsValid() == True:
        print("RVC Camera is valid!")
    else:
        print("RVC Camera is not valid!")
        RVC.X1.Destroy(x)
        RVC.SystemShutdown()
        return 1

    # Open RVC Camera.
    ret1 = x.Open()

    # Test RVC Camera is opened or not.
    if x.IsOpen() == True:
        print("RVC Camera is opened!")
    else:
        print("Failed to open camera! Please check whether the camera is connected and make sure it is not occupied and supports X1.")
        RVC.X1.Destroy(x)
        RVC.SystemShutdown()
        return 1
    
    # Print ExposureTime Range
    _, exp_range_min, exp_range_max = x.GetExposureTimeRange()
    print("ExposureTime Range:[{}, {}]".format(exp_range_min, exp_range_max))

    # Save CaptureOptions
    x1_save_opts = RVC.X1_CaptureOptions()
    x1_save_opts.exposure_time_3d = 20
    x1_save_opts.exposure_time_2d = 21
    x.SaveCaptureOptionParameters(x1_save_opts)

    # Load CaptureOptions
    ret, x1_opts= x.LoadCaptureOptionParameters()
    PrintCaptureOptionX1(x1_opts)

    # Close RVC Camera.
    x.Close()

    # Destroy RVC Camera.
    RVC.X1.Destroy(x)

    # Shutdown RVC System.
    RVC.SystemShutdown()

    return 0


def UseX2():
    # Initialize RVC system.
    RVC.SystemInit()

    # Choose RVC Camera type (USB, GigE or All)
    opt = RVC.SystemListDeviceTypeEnum.All

    # Scan all RVC Camera devices.
    ret, devices = RVC.SystemListDevices(opt)
    print("RVC Camera devices number:%d" % len(devices))

    # Find whether any RVC Camera is connected or not.
    if len(devices) == 0:
        print("Can not find any RVC Camera!")
        RVC.SystemShutdown()
        return 1
    print("devices size = %d" % len(devices))
    if devices[0].IsFirmwareMatch() == False:
        print("device firmware mismatch, Please use RVCManager to upgrade the firmware")
        RVC.SystemShutdown()
        return 1
    # Create a RVC Camera
    x = RVC.X2.Create(devices[0])

    # Test RVC Camera is valid or not
    if x.IsValid() == False:
        print("RVC Camera is not valid!")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(1)
        
    #PrintCaptureMode(devices[0])

    # Open RVC Camera
    ret1 = x.Open()

    # Test RVC Camera is opened or not
    if x.IsOpen() == False:
        print("Failed to open camera! Please check whether the camera is connected and make sure it is not occupied and supports X2.")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(1)

    ret, info = devices[0].GetDeviceInfo()
    camera_id = RVC.CameraID_Left
    if info.support_extra:
        camera_id = RVC.CameraID_Extra
        
    # Set capture parameters
    cap_opt = RVC.X2_CaptureOptions()
    # Transform point map's coordinate to camera
    cap_opt.transform_to_camera = camera_id
    # Set camera exposure time (3~100) ms
    cap_opt.exposure_time_2d = 20
    cap_opt.exposure_time_3d = 20
    # range in [0, 10]. the default value is 3. The contrast of point less than this value will be treat * as invalid point and be removed.
    cap_opt.light_contrast_threshold = 2
    # Set projector color, defult is RVC.ProjectorColor_Blue
    cap_opt.projector_color = RVC.ProjectorColor_Blue
    x.SaveCaptureOptionParameters(cap_opt)

    # Load CaptureOptions
    ret, x2_opts= x.LoadCaptureOptionParameters()
    PrintCaptureOptionX2(x2_opts)

    # Close RVC Camera.
    x.Close()

    # Destroy RVC Camera.
    RVC.X2.Destroy(x)

    # Shutdown RVC System.
    RVC.SystemShutdown()

    return 0

if __name__ == "__main__":
    UseX1()
    UseX2()
